frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java

```diff
@@ -608,7 +608,13 @@ private ContainerLaunchContext setupContainerLaunchContext(TaskStatus taskStatus
     localEnvs.put(GlobalConstants.ENV_VAR_CONTAINER_GPUS, taskStatus.getContainerGpus().toString());
 
     String containerPortsString = ValueRangeUtils.toEnviromentVariableString(portRanges, portDefinitions);
-    localEnvs.put(GlobalConstants.ENV_VAR_CONTAINER_PORTS, containerPortsString);
+    if (portDefinitions != null && !portDefinitions.isEmpty()) {
+      if (containerPortsString.split(";").length == portDefinitions.size()) {
+        localEnvs.put(GlobalConstants.ENV_VAR_CONTAINER_PORTS, containerPortsString);
+      } else {
+        throw new NotAvailableException("ENV_VAR_CONTAINER_PORTS:" + containerPortsString + " is not meet the request in portDefinitions");
+      }
```

Comment:
This logic should not be in setupContainerLaunchContext and should be in ports selection sections

Reply:
And outside will not catch the NotAvailableException, it will cause the whole AM crash.

---
In reply to: [182636593](https://github.com/Microsoft/pai/pull/495#discussion_r182636593) [](ancestors = 182636593)

Reply:
We need to know in which case containerPortsString.split(";").length != portDefinitions.size().
If it is expected, we should test in testContainer

---
In reply to: [182636846](https://github.com/Microsoft/pai/pull/495#discussion_r182636846) [](ancestors = 182636846,182636593)

Reply:
Not need to throw exception. We just need to LogError this issue. it is not due to resource-not-available as it has been checked before.

---
In reply to: [182637566](https://github.com/Microsoft/pai/pull/495#discussion_r182637566) [](ancestors = 182637566,182636846,182636593)